### PR TITLE
MAINT: using common test primitive for banned primitives check in locations

### PR DESCRIPTION
**Pull request by samir-nasibli**
### Description
Suggested changes:
* Generalized for further use of checker for banned primitives

Discussed here https://github.com/intel/scikit-learn-intelex/pull/1839#issuecomment-2156245612 

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes, if necessary (updated in # - _add PR number_)
- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [x] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

**Review by icfaust — COMMENTED** [review]
I find the justification for the PR to be weak, but since the work has been done, I ask that you show that fuctionality hasn't been reduced. please show the CI failing for private CI and public CI for target_offload and sklearn_check_version when they are placed in the codebase. ideally by temporarily modifying your PR, running CIs and then returning them back.

**Comment by samir-nasibli** [comment]
@icfaust Thank you for the feedback!
Can you please clarify what exactly was missing for justification, I will try to expand the PR description.
Yeah, it seem I should update this test runners.
Btw current internal CI fail is not related to the changes.

**Comment by icfaust** [comment]
> @icfaust Thank you for the feedback! Can you please clarify what exactly was missing for justification, I will try to expand the PR description. Yeah, it seem I should update this test runners. Btw current internal CI fail is not related to the changes.

The justification of this PR is that this is a movement since it is a developer-only feature. However, all tests are included in the install package, and those are for developers and not end users. That is also the case with sklearn, where we install sklearn and then directly use the testing suite as developers, even those it is in the user install. 

I would like to see the tests throw a failure (to show they are working in CI). Could you add cases of sklearn_check_version and device_offload to show that the tests properly fail in CI? We can remove them later before merging.

**Comment by samir-nasibli** [comment]
> > @icfaust Thank you for the feedback! Can you please clarify what exactly was missing for justification, I will try to expand the PR description. Yeah, it seem I should update this test runners. Btw current internal CI fail is not related to the changes.
> 
> The justification of this PR is that this is a movement since it is a developer-only feature. However, all tests are included in the install package, and those are for developers and not end users. That is also the case with sklearn, where we install sklearn and then directly use the testing suite as developers, even those it is in the user install.
> 
> I would like to see the tests throw a failure (to show they are working in CI). Could you add cases of sklearn_check_version and device_offload to show that the tests properly fail in CI? We can remove them later before merging.

Thank you @icfaust !

I thought about it once again and I think that your opinion is reseanoble and it really might be worth it to include this in standard testing of the package.
But I wouldn't want to turn off these tests. I would suggest muting cases if there are any disabled or exceptions.
In addition, I want to see how well these tests work, and whether they affect the validation process in our CI

File: sklearnex/tests/test_common.py [hunk 1]
```diff
@@ -21,15 +21,12 @@
 import re
 import sys
 import trace
-from glob import glob
 
-import numpy as np
 import pytest
-import scipy
-import sklearn.utils.validation
```
Comment (samir-nasibli) [inline sklearnex/tests/test_common.py:29]: removed unused imports

**Comment by samir-nasibli** [comment]
@icfaust I have returned back tests, but generalized in a test primitive common parts to avoid code duplication.

File: onedal/tests/test_common.py [hunk 2]
```diff
@@ -14,27 +14,43 @@
...
 from glob import glob
 
-import pytest
 
-
-def test_sklearn_check_version_ban():
-    """This test blocks the use of sklearn_check_version
-    in onedal files. The versioning should occur in the
-    sklearnex package for clarity and maintainability.
+def _test_primitive_usage_ban(primimtive_name, banned_locations, allowed_locations=None):
```
Comment (icfaust) [inline onedal/tests/test_common.py:22]:
```suggestion
def _check_primitive_usage_ban(primitive_name, package, allowed_locations=None):
```
I know the underscore will avoid pytest including it as a test, but I think changing 'test' to 'check' would be nice.  I think its best also if a single package is checked instead of 'banned_locations'. Also fix a spelling mistake.

File: onedal/tests/test_common.py [hunk 3]
```diff
@@ -14,27 +14,43 @@
 # limitations under the License.
 # ==============================================================================
 
+import importlib
```
Comment (icfaust) [inline onedal/tests/test_common.py:17]:
```suggestion
import pkgutil
```

File: onedal/tests/test_common.py [hunk 4]
```diff
@@ -14,27 +14,43 @@
...
-    """This test blocks the use of sklearn_check_version
-    in onedal files. The versioning should occur in the
-    sklearnex package for clarity and maintainability.
+def _test_primitive_usage_ban(primimtive_name, banned_locations, allowed_locations=None):
+    """This test blocks the usage of the primitive in
+    in certain files.
     """
-    from onedal import __file__ as loc
+
+    loc = importlib.import_module(banned_locations).__file__
```
Comment (icfaust) [inline onedal/tests/test_common.py:27]:
```suggestion
    loc = pkgutil.get_loader(package).filename
```
This removes the need to import the package, probably something that I should have done in the original implementation.

File: onedal/tests/test_common.py [hunk 5]
```diff
@@ -14,27 +14,43 @@
...
     path = loc.replace("__init__.py", "")
     files = [y for x in os.walk(path) for y in glob(os.path.join(x[0], "*.py"))]
 
     output = []
 
     for f in files:
-        if open(f, "r").read().find("sklearn_check_version") != -1:
-            output += [f.replace(path, "onedal" + os.sep)]
+        if open(f, "r").read().find(primimtive_name) != -1:
+            output += [f.replace(path, banned_locations + os.sep)]
```
Comment (icfaust) [inline onedal/tests/test_common.py:36]:
```suggestion
        if open(f, "r").read().find(primitive_name) != -1:
            output += [f.replace(path, package + os.sep)]
```

File: onedal/tests/test_common.py [hunk 6]
```diff
@@ -14,27 +14,43 @@
...
+
+
+def test_sklearn_check_version_ban():
+    """This test blocks the use of sklearn_check_version
+    in onedal files. The versioning should occur in the
+    sklearnex package for clarity and maintainability.
+    """
+    output = _test_primitive_usage_ban(
+        primimtive_name="sklearn_check_version", banned_locations="onedal"
+    )
```
Comment (icfaust) [inline onedal/tests/test_common.py:53]:
```suggestion
    output = _check_primitive_usage_ban(
        primitive_name="sklearn_check_version", package="onedal"
    )
```

File: sklearnex/tests/test_common.py [hunk 7]
```diff
@@ -109,23 +106,13 @@ def test_target_offload_ban():
...
-
-    # remove this file from the list
-    for allowed in ALLOWED_LOCATIONS:
-        output = [i for i in output if allowed not in i]
-
+    output = _test_primitive_usage_ban(
+        primimtive_name="target_offload",
+        banned_locations="sklearnex",
+        allowed_locations=TARGET_OFFLOAD_ALLOWED_LOCATIONS,
+    )
```
Comment (icfaust) [inline sklearnex/tests/test_common.py:113]:
```suggestion
    output = _check_primitive_usage_ban(
        primitive_name="target_offload",
        package="sklearnex",
        allowed_locations=TARGET_OFFLOAD_ALLOWED_LOCATIONS,
    )
```

File: sklearnex/tests/test_common.py [hunk 8]
```diff
@@ -21,15 +21,12 @@
...
-from glob import glob
 
-import numpy as np
 import pytest
-import scipy
-import sklearn.utils.validation
 from sklearn.utils import all_estimators
 
 from daal4py.sklearn._utils import sklearn_check_version
+from onedal.tests.test_common import _test_primitive_usage_ban
```
Comment (icfaust) [inline sklearnex/tests/test_common.py:29]:
```suggestion
from onedal.tests.test_common import _check_primitive_usage_ban
```

File: sklearnex/tests/test_common.py [hunk 9]
```diff
@@ -109,23 +106,13 @@ def test_target_offload_ban():
...
-        output = [i for i in output if allowed not in i]
-
+    output = _test_primitive_usage_ban(
+        primimtive_name="target_offload",
+        banned_locations="sklearnex",
+        allowed_locations=TARGET_OFFLOAD_ALLOWED_LOCATIONS,
+    )
     output = "\n".join(output)
-    assert output == "", f"sklearn versioning is occuring in: \n{output}"
+    assert output == "", f"target offloading is occuring in: \n{output}"
```
Comment (icfaust) [inline sklearnex/tests/test_common.py:115]: good catch

**Review by icfaust — COMMENTED** [review]
Big things are: spelling mistakes on 'primitive', using pkgutil instead of importlib (also in standard library), and a stylistic change from "_test" to "_check". Thanks also for catching some mistakes in the original implementation. These changes and we can merge quickly.

File: onedal/tests/test_common.py [hunk 10]
```diff
@@ -15,26 +15,46 @@
 # ==============================================================================
 
 import os
+import pkgutil
```
Comment (icfaust) [inline onedal/tests/test_common.py:18]:
```suggestion
import importlib
```

File: onedal/tests/test_common.py [hunk 11]
```diff
@@ -15,26 +15,46 @@
...
+    """This test blocks the usage of the primitive in
+    in certain files.
     """
-    from onedal import __file__ as loc
+
+    # TODO:
+    # Address deprecation warning.
+    # The function "get_loader" is deprecated Use importlib.util.find_spec() instead.
+    # Will be removed in Python 3.14.
+    loc = pkgutil.get_loader(package).get_filename()
```
Comment (icfaust) [inline onedal/tests/test_common.py:31]:
```suggestion
    loc = importlib.util.find_spec(package).origin
```
I'm sorry to have made the previous suggestion. This should solve the todo, and thank you very much for putting in the effort on the changes.

**Review by icfaust — COMMENTED** [review]
Sorry again for making you change this again.  This should now remove the added TODO.

**Comment by samir-nasibli** [comment]
> Sorry again for making you change this again. This should now remove the added TODO.

No problem. I tought we can update it later, but maybe make sense do it now. Applied suggestions. Thank you!

**Review by icfaust — APPROVED** [review]
Let the CI run through. Merge ready assuming green
